File: docs-source/docs/modules/shopping-microservices/examples/01-shopping-cart-service-scala/src/main/scala/shopping/cart/ShoppingCartServiceImpl.scala
package shopping.cart

import scala.concurrent.Future

import org.slf4j.LoggerFactory

class ShoppingCartServiceImpl extends proto.ShoppingCartService {

  private val logger = LoggerFactory.getLogger(getClass)

  override def addItem(in: proto.AddItemRequest): Future[proto.Cart] = { // <1>
    logger.info("addItem {} to cart {}", in.itemId, in.cartId)
    Future.successful(proto.Cart(items = List(proto.Item(in.itemId, in.quantity))))
  }

}

File: docs-source/docs/modules/microservices-tutorial/examples/01-shopping-cart-service-java/src/main/java/shopping/cart/ShoppingCartServiceImpl.java
package shopping.cart;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import shopping.cart.proto.AddItemRequest;
import shopping.cart.proto.Cart;
import shopping.cart.proto.Item;
import shopping.cart.proto.ShoppingCartService;

public final class ShoppingCartServiceImpl implements ShoppingCartService {

  private final Logger logger = LoggerFactory.getLogger(getClass());

  @Override
  public CompletionStage<Cart> addItem(AddItemRequest in) { // <1>
    logger.info("addItem {} to cart {}", in.getItemId(), in.getCartId());
    Item item = Item.newBuilder().setItemId(in.getItemId()).setQuantity(in.getQuantity()).build();
    Cart cart = Cart.newBuilder().addItems(item).build();
    return CompletableFuture.completedFuture(cart);
  }
}
